xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authentication/xwiki-platform-security-authentication-default/src/main/java/org/xwiki/security/authentication/internal/ResetPasswordMailSender.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.security.authentication.internal;
+
+import java.net.URL;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.mail.MessagingException;
+import javax.mail.internet.InternetAddress;
+import javax.mail.internet.MimeMessage;
+
+import org.apache.commons.lang3.StringUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.localization.ContextualLocalizationManager;
+import org.xwiki.mail.MailListener;
+import org.xwiki.mail.MailSender;
+import org.xwiki.mail.MailSenderConfiguration;
+import org.xwiki.mail.MailStatus;
+import org.xwiki.mail.MailStatusResult;
+import org.xwiki.mail.MimeMessageFactory;
+import org.xwiki.mail.SessionFactory;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.security.authentication.api.ResetPasswordException;
+
+import com.xpn.xwiki.XWikiContext;
+
+/**
+ * Component dedicated to send reset password information by email.
+ *
+ * @version $Id$
+ * @since 13.1RC1
+ */
+@Component(roles = ResetPasswordMailSender.class)
```

Comment:
I'm personally not fond of this. I much prefer a clean interface. Otherwise if you're not careful you end up adding new method that shouldn't be APIs. I know this is internal but still.

Reply:
Well for this one I really do prefer keeping it like that and not exposing it as an interface since there's no real need of using it externally. At first it was even implemented as part of the DefaultResetPasswordManager, I put it outside because it was becoming really too big.

Reply:
> Otherwise if you're not careful you end up adding new method that shouldn't be APIs

As you said since it's internal, so it's not an API. I really don't see the point of having an interface/implementation for something like this, it's painful and does not bring any kind of value.